List View: Scroll selected block into view when single block selection changes

andrewserong opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

Fixes #46828, fixes #30432

While the list view is open, scroll the selected item into view if it is not already in view.

A caveat on the approach in this PR: I'm not sure how performant it is, so very happy for any feedback on this one!

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

As described in #46828, with the list view open and a long post or page, it can be confusing if you change the selection within the editor canvas, but the list view does not scroll that block into view. It can be quite easy to lose your place on longer posts.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

* Add a hook to the list view that scrolls the selected block's element into view if it isn't already visible.
* Use this at the `ListViewBlock` level, and also at the `ListViewBranch` level to factor in very long posts/pages where the placeholder for the selected block falls outside of the current window where "real" list view blocks are rendered. This way, if a user selects a block that is outside of the currently windowed area, we can still scroll it into view.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->

In a really long post or page, with the list view opened, scroll the editor canvas to far down the page or post and select a block outside of the currently visible area within the list view. The list view should scroll the selected block into view.

To test the windowed area versus the non-windowed area, add a very large number of blocks (e.g. 60+) and try selecting from the very top block to the very bottom block in the list, via the editor canvas.

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

It should be possible to test this via keyboard – with the list view opened and focus on the editor canvas, press the down arrow to navigate down blocks, and as you scroll past the bottom of the screen, or past the top of the screen, the list view should always display the selected block.

## Screenshots or screencast <!-- if applicable -->

In the below screengrab, with the list view open, we go from the first block being selected, to scrolling the editor canvas and selecting a block that is outside of the visible area of the list view. When the block is selected, the list view is scrolled so that the selected block is at the top of the visible area of the list view.

https://user-images.githubusercontent.com/14988353/210718105-0f735197-31f7-461d-ad5c-114a36c88d94.mp4

andrewserong committed:
List View: Try scrolling selected blocks into view when single block selection changes

apeatling commented on packages/block-editor/src/components/list-view/use-list-view-scroll-into-view.js:
```diff
@@ -0,0 +1,33 @@
+		const { innerHeight } = window;
+
+		// If the selected block is not visible, scroll to it.
+		// The hard-coded value of 110 corresponds to the position at the top of the scrollable area.
```
What does the "position at the top of the scrollable area" mean here? Is that the height of the UI at the top?

andrewserong replied:
> Is that the height of the UI at the top?

Yep! Thanks for the question here — I couldn't quite work out how to describe it. The `top` value is from the top of the screen, but we want to calculate the value in relation to the visible area which is up to the bottom of this tabbed area:

![image](https://user-images.githubusercontent.com/14988353/210908471-94d988bb-1e80-412f-ba61-84a210745ab5.png)

We can get that value by doing a look up using `getScrollContainer` in the `@wordpress/dom` package, but I included a hard-coded value for now, since I think that might be a bit more performant than having to do a look up. Happy to swap it out for the "real" calculation if you think that'd be a better approach (or to try to update the comment to be clearer!)

talldan replied:
I definitely think it'd be better to use `getScrollContainer`, especially as this UI has undergone recent change, and might end up being changed further.

apeatling replied:
Thanks, I thought there might be a solution that wouldn't cause problems if the UI changed. 👍

talldan commented on packages/block-editor/src/components/list-view/branch.js:
```diff
@@ -116,6 +117,16 @@ function ListViewBranch( props ) {
+	// The "real" selected item is also observed in ListViewBlock, when rendered.
+	const invisibleSelectedItemRef = useRef();
+	useListViewScrollIntoView( {
+		firstSelectedBlockClientId: selectedClientIds?.[ 0 ],
+		numBlocksSelected: selectedClientIds?.length,
```
The interface could potentially be simplified to just `useListViewScrollIntoView( { selectedClientIds, selectedItemRef } )`, moving the logic that grabs the first clientId and the length into the hook.

talldan commented on packages/block-editor/src/components/list-view/branch.js:
```diff
@@ -116,6 +117,16 @@ function ListViewBranch( props ) {
+	// For long lists where the selected item may fall outside of the current window,
+	// pass a reference to the corresponding placeholder row for the selected item.
+	// The "real" selected item is also observed in ListViewBlock, when rendered.
+	const invisibleSelectedItemRef = useRef();
+	useListViewScrollIntoView( {
```
Is it possible to use the hook only here and remove the usage in `ListViewBlock`?

I find the naming `invisibleSelectedItemRef` a bit confusing, and it might be cleaner if there were only one `selectedBlockRef` that were passed to either the placeholder row or `ListViewBlock` as a prop.

talldan commented on packages/block-editor/src/components/list-view/use-list-view-scroll-into-view.js:
```diff
@@ -0,0 +1,33 @@
+} ) {
+	useEffect( () => {
+		if (
+			numBlocksSelected !== 1 ||
```
What's the reason for the early return when there's a multiselection?

andrewserong replied:
This was mostly to try to simplify the approach — when a multiselection is applied, we can't know (in this implementation) whether or not it was done from the list view, in which case we wouldn't want the list view to be scrolled. For example, if a user selects a block, scrolls down the list view and then shift-selects a block further down, we don't want the list view to suddenly scroll back up to the initially selected block.

I imagined the main use case for scrolling the list view is changing single block selection, but happy to dig in further if you think we should handle multiselection from the outset.

talldan replied:
Ah, cool, that does sound like a good reason to start with single block selection. It'd be good to add a comment to that effect.

talldan commented on packages/block-editor/src/components/list-view/use-list-view-scroll-into-view.js:
```diff
@@ -0,0 +1,33 @@
+	}, [
+		firstSelectedBlockClientId,
+		numBlocksSelected,
+		selectedItemRef?.current,
```
A ref changing won't trigger a re-render, so I don't think this needs to be specified.

talldan reviewed (commented):
This is very cool, thanks for working on it!

talldan commented on packages/block-editor/src/components/list-view/block.js:
```diff
@@ -220,6 +221,14 @@ function ListViewBlock( {
+	useListViewScrollIntoView( {
+		firstSelectedBlockClientId: selectedClientIds?.[ 0 ],
+		numBlocksSelected: selectedClientIds?.length,
+		selectedItemRef: isSelected ? cellRef : undefined,
```
There's a small inconsistency in that in one place the ref is bound to a row, and in the other a cell. I think it'd be tidied up by only calling `useListViewScrollIntoView` in `ListViewBranch` (removing it here) and having `ListViewBlock` accept a ref to its row.

andrewserong commented:
Great feedback, thanks for the detailed review, Dan! I ran out of time today, but I'll dig through the rest of the feedback and update this PR hopefully later on this week 🙂

andrewserong committed:
Use getScrollContainer and calculate real top position of scrollable area instead of using a hard-coded value

andrewserong commented:
I've removed the hard-coded top position of the scroll container and replaced with `getScrollContainer` to use an accurate measurement. This PR isn't quite ready for a re-review (I still need to tidy up the refs, etc), so I'll finish off this pass of changes next week.

andrewserong committed:
Try rearranging things so that the ref is always attached at the row level

andrewserong commented:
Update: in https://github.com/WordPress/gutenberg/pull/46895/commits/be7edb2a4495434f442d874f0cbea7589ddf57cc I've moved things around a bit. Some of the issues I ran into while messing around with the refs:

Using a ref at the branch level seemed to cause reliability issues, as conditionally setting the ref between the real selected item or the placeholder row caused an issue once we removed it from the block level and only attached it at the branch level. I think this might have been either due to the async rendering via the async provider (I'm not too sure on the details there), or because changing the ref value doesn't cause a re-render. The result was that sometimes if you clicked from far down in the document back to the top of the document, the `useListViewScrollIntoView` would receive a null reference for the `ref` rather than the "real" node, even though the ref had updated in the parent component.

I _think_ a more explicit way to do it might be to ensure that the hook is always attached, but setting it at the row level. So far I've done this by doing the following:

* Allow passing a ref to `ListViewLeaf`
* In `ListViewBlock` pass a `ref` to the `ListViewLeaf` and retain usage of `useListViewScrollIntoView`
* In `ListViewBranch` remove the hook from the parent level, and conditionally render a `PlaceholderRow` for the selected block if it's to be rendered as a placeholder. Within this component, use `useListViewScrollIntoView`.

With the above changes, we're no longer trying to wrestle with the `ref`, and it also avoids having to store the reference to the DOM node in state. On the other hand, it still depends on multiple calls to `useListViewScrollIntoView`.

I'm not sure I'm entirely sold on the approach I've gone with here, but it's the one that feels the most stable in usage to me, so far. Very happy for feedback on this if and when folks have time — I'm quite happy to go in another direction with this, as the current one doesn't feel quite neat enough for my own liking, but I wasn't too sure how else to achieve it while avoiding the issue with conditionally attaching `ref` at the branch level 🤔

talldan commented on packages/block-editor/src/components/list-view/branch.js:
```diff
@@ -186,11 +205,18 @@ function ListViewBranch( props ) {
-							</tr>
-						) }
+						{ ! showBlock &&
+							( isSelected ? (
+								<PlaceholderRow
+									clientId={ clientId }
+									isSelected={ isSelected }
+									selectedClientIds={ selectedClientIds }
+								/>
+							) : (
+								<tr>
+									<td className="block-editor-list-view-placeholder" />
+								</tr>
+							) ) }
```
I don't quite follow the reasoning for the ternary here, as the else state seems to render the same thing as `PlaceholderRow`.

andrewserong replied:
I wasn't sure if there was a performance reason for using `<tr>` directly over a separate React component for the placeholder rows, so this ternary preserved the `<tr>` for the majority of the placeholders. It'd be neater to remove it and just use `PlaceholderRow`, though, so I'll have a play with consolidating and see how that goes.

talldan reviewed (commented):
> I think this might have been either due to the async rendering via the async provider

Yep, that's probably it. The problem is likely that if you call the hook in `ListViewBranch`, the hook will be outside the `<AsyncModeProvider value={ false}>`, so any data updates will be async instead of sync, and may not trigger a timely re-render.

`useListViewScrollIntoView` needs to be defined in one of the children of the async mode provider to solve that, which seems to match what the latest commits do.

The idea of making the Placeholder its own component seems a good one and makes things more consistent, but I think it could be in a separate file with some docs to explain its purpose. Or the logic is moved into `ListViewBlock` or `ListViewLeaf`.

andrewserong commented:
> The idea of making the Placeholder its own component seems a good one and makes things more consistent, but I think it could be in a separate file with some docs to explain its purpose. Or the logic is moved into ListViewBlock or ListViewLeaf.

Sounds good! Thanks for the feedback, Dan — I was a little on the fence about the approach, but I think this will help tidy it up 👍

andrewserong committed:
Move placeholder to its own file

andrewserong committed:
Tidy up a little

andrewserong committed:
Tidy comments

andrewserong commented:
I've moved the placeholder into a separate file and renamed it `ListViewLeafPlaceholder` for consistency with the other component naming. I think this PR's feeling a lot neater than it was earlier, so I'm feeling pretty optimistic about the direction now, thanks again for all the feedback/pointers.

Let me know if anyone would like any other changes made, and I can keep tinkering! It'd also be good to double-check the performance stats once the performance job finishes, to make sure this doesn't introduce any regressions there.

apeatling reviewed (commented):
Gave this another run through and it's looking and working well. I checked the performance tests job and noticed some large changes on the site editor stats. Looks unrelated, but worth running again?

andrewserong commented:
Thanks for taking this for another spin @apeatling! Yeah, in theory I wouldn't think this would affect performance like that, so worth running it again. I'll give this PR a rebase, then we can see what the results are looking like. For posterity, the current performance results as of the last commit appear to be:

```
┌──────────────────────┬──────────────────────────────────────────┬─────────────┐
│       (index)        │ 5d2c8647bf92d4b7b4c52e699b9d9528bc057043 │    trunk    │
├──────────────────────┼──────────────────────────────────────────┼─────────────┤
│    serverResponse    │                '88.63 ms'                │ '126.43 ms' │
│      firstPaint      │               '465.97 ms'                │ '347.3 ms'  │
│   domContentLoaded   │               '939.93 ms'                │ '495.3 ms'  │
│        loaded        │                '943.2 ms'                │ '495.93 ms' │
│ firstContentfulPaint │               '1296.43 ms'               │ '595.5 ms'  │
│      firstBlock      │               '9749.5 ms'                │ '1632.4 ms' │
│         type         │                '48.95 ms'                │ '43.32 ms'  │
│       minType        │                '42.3 ms'                 │ '40.81 ms'  │
│       maxType        │               '288.06 ms'                │ '107.88 ms' │
└──────────────────────┴──────────────────────────────────────────┴─────────────┘
```

andrewserong commented:
Update: after a rebase, I'm still seeing `firstBlock` time in the site editor as being much slower than on `trunk` in the performance tests results. After double checking another couple of open PRs (https://github.com/WordPress/gutenberg/pull/47389, https://github.com/WordPress/gutenberg/pull/47384) it seems this is pretty consistent with that Github Action running on open PRs, so I suspect it isn't caused by this PR. It's a bit unfortunate to see, though!

kevin940726 commented on packages/block-editor/src/components/list-view/use-list-view-scroll-into-view.js:
```diff
@@ -0,0 +1,53 @@
+		if (
+			! isSelected ||
+			numBlocksSelected !== 1 ||
+			! rowItemRef?.current ||
```
`ref` should always be present, if it's not then I think we will have a bigger issue and we want to catch that. 😅 

```suggestion
			! rowItemRef.current ||
```

kevin940726 commented on packages/block-editor/src/components/list-view/use-list-view-scroll-into-view.js:
```diff
@@ -0,0 +1,53 @@
+			! isSelected ||
+			numBlocksSelected !== 1 ||
+			! rowItemRef?.current ||
+			! rowItemRef.current.scrollIntoView
```
Not sure why this is needed? `scrollIntoView` should be available for all browsers that we support.

andrewserong replied:
Great point, I'd seen a check for it elsewhere in GB but that's probably from a long time ago and doesn't appear to be needed now 👍

kevin940726 commented on packages/block-editor/src/components/list-view/use-list-view-scroll-into-view.js:
```diff
@@ -0,0 +1,53 @@
+		const { innerHeight } = defaultView;
+
+		// If the selected block is not currently visible, scroll to it.
+		if ( top < topOfScrollContainer || top + height > innerHeight ) {
```
This works for the list view, but I think it is by accident. Instead of checking for the window's inner height, we should check for the `scrollContainer`'s height.

The current implementation also has a bit of inconsistency. If you click on a row that has its top half covered by the container, then the row will be scrolled to show the full height in the viewport. This is a nice improvement, and I think it should be the expected behavior. However, if the same happens for a row that has its bottom half covered by the container, clicking it won't scroll the container.

Perhaps, we can refactor this into something like this for simplicity:

```js
const rowRect = rowItemRef.current.getBoundingClientRect();
const scrollContainerRect = scrollContainer.getBoundingClientRect();

if ( rowRect.top < scrollContainerRect.top || rowRect.bottom > scrollContainerRect.bottom ) {
  rowItemRef.current.scrollIntoView();
}
```

andrewserong replied:
Ah, yes, that logic's much better, thank you! I think my previous logic was a bit of a mix between two different approaches I'd hacked together but that's much neater and more reliable 🙂

kevin940726 commented on packages/block-editor/src/components/list-view/use-list-view-scroll-into-view.js:
```diff
@@ -0,0 +1,53 @@
+		if ( top < topOfScrollContainer || top + height > innerHeight ) {
+			rowItemRef.current.scrollIntoView();
+		}
+	}, [ isSelected, numBlocksSelected ] );
```
`rowItemRef` should also be listed in the dependencies (because the hook doesn't know if it's a ref or not). :)

```suggestion
	}, [ isSelected, numBlocksSelected, rowItemRef ] );
```

andrewserong replied:
Added 👍

kevin940726 commented on packages/block-editor/src/components/list-view/use-list-view-scroll-into-view.js:
```diff
@@ -0,0 +1,53 @@
+	selectedClientIds,
+	rowItemRef,
+} ) {
+	const numBlocksSelected = selectedClientIds?.length;
```
Nit: Let's rename this variable into something like `isMultiSelected` so that we don't have to needlessly rerun the effect when the number of selected blocks changes.

andrewserong replied:
Good call. I changed the name to `isSingleSelection` because I think that gets slightly closer to the existing check, but either name would work 🙂

andrewserong commented:
Thanks for the detailed review @kevin940726, much appreciated! 🙇
I'm a little stretched for time today, so if I don't get to implementing your suggestions today (they all look like good notes 👍), I'll follow up on Monday 🙂

andrewserong committed:
Remove unneeded optional chaining

Co-authored-by: Kai Hao <kevin830726@gmail.com>

andrewserong committed:
Simplify and improve logic based on feedback

Co-authored-by: Kai Hao <kevin830726@gmail.com>

kevin940726 commented on packages/block-editor/src/components/list-view/use-list-view-scroll-into-view.js:
```diff
@@ -0,0 +1,48 @@
+	selectedClientIds,
+	rowItemRef,
+} ) {
+	const isSingleSelection = selectedClientIds?.length === 1;
```
Nit: Is `selectedClientIds` required in this hook? If it is then we probably need the nullish check here.

andrewserong replied:
What did you have in mind for this one? Would it be something like the following:

```
const isSingleSelection = selectedClientIds ?? selectedClientIds?.length === 1;
```

If so, I'm not sure if it's needed, because it doesn't appear to cause any errors if `selectedClientIds` isn't passed in, since the destructuring assignment for `selectedClientIds` means that the optional chaining doesn't appear to throw an error because is `selectedClientIds` is defined as undefined, rather than being truly undefined... I think 😅

For example, I tried the following in the console:

<img width="491" alt="image" src="https://user-images.githubusercontent.com/14988353/218363304-18916638-6679-40b0-8787-dce5e4f3df95.png">

Was there another case you were thinking of, though, where adding a nullish check would improve things? Happy to keep tweaking this!

Also thanks for your comment, TIL that optional chaining doesn't perform a check for us to make sure that the variable is defined before attempting to access the property, so I'm very happy learn more here, too 🙂

kevin940726 replied:
Oh, oops! I mean optional chaining, not nullish operator 😅 .

```suggestion
	const isSingleSelection = selectedClientIds.length === 1;
```

Sorry for the misunderstanding!

andrewserong replied:
Ah, excellent, thanks for clarifying! Yes, `selectedClientIds` should always be both required and an array, from looking at [the selector](https://github.com/WordPress/gutenberg/blob/f50ab9815d6b8ae00781a7db8c31e893af0ac19f/packages/block-editor/src/store/selectors.js#L710), it'll always return an array, even if it's empty, so this looks safe to me 👍

kevin940726 reviewed (approved):
LGTM 👍 , great work!

andrewserong committed:
Remove unneeded optional chaining

Co-authored-by: Kai Hao <kevin830726@gmail.com>

andrewserong commented:
After giving this a re-test, I think this PR currently introduces a regression when scrolling a very long list quickly — the list snaps back up to the top unexpectedly. It's possibly due to the switching between list view items being rendered as a placeholder versus a "real" row item — my hunch is that when the component is swapped, the hook gets fired thinking that the block has been freshly selected, or something along those lines... here's a screengrab:

![2023-02-13 17 26 09](https://user-images.githubusercontent.com/14988353/218386878-26b9f873-295d-4fda-b7cb-c40ad096f44c.gif)

I might need to give the hook logic a little more thought — we only want it to fire when a selection change has been made 🤔

andrewserong committed:
Revert placeholder component, update showBlock logic so that selected blocks are rendered as real ListViewBlock components

andrewserong commented:
Update: I believe I've fixed the regression I encountered yesterday (https://github.com/WordPress/gutenberg/pull/46895#issuecomment-1427420229) by doing the following:

* Update the `showBlock` logic so that if a block is selected, we render a real `ListViewBlock` component instead of the placeholder.
* Revert the `ListViewLeafPlaceholder` component, and restore using a simple `<tr><td>` element, thereby removing the duplicate logic calling the `useListViewScrollIntoView` hook. While I liked the placeholder component, it should be simpler using DOM elements here rather than introducing more React components 🤞

The result is that now there's only one "version" of a selected block in the list view, so the hook's `scrollIntoView()` call now appears to fire correctly when scrolling a very long list view really fast — the swapping between a placeholder and a real component no longer occurs for selected blocks.

I _think_ this is working pretty well now, but could probably use another review / re-test to confirm that I haven't missed anything if anyone gets a chance 🙂

Thanks everyone for all the reviews and help so far! 🙇

kevin940726 reviewed (approved):
Works great in my testing. Thanks!

apeatling reviewed (approved):
Given this one another good run through and everything is working as expected. I could not reproduce the list jumping issue you found, so that one looks good to me. :+1:

andrewserong commented:
Thanks for the re-reviews! Much appreciated 🙇

ntsekouras commented:
I just cherry-picked this PR to the wp/6.2 branch to get it included in the next release: 7951add272
